backends/gstreamer/lib.rs

```diff
@@ -58,36 +61,90 @@ use servo_media_player::{Player, PlayerEvent, StreamType};
 }
 
 pub struct GStreamerBackend {
     capture_mocking: AtomicBool,
+    muteables: Mutex<HashMap<ClientContextId, Vec<(usize, Weak<Mutex<dyn Muteable>>)>>>,
+    next_muteable_id: AtomicUsize
 }
 
 impl Backend for GStreamerBackend {
     fn create_player(
         &self,
+        id: &ClientContextId,
         stream_type: StreamType,
         sender: IpcSender<PlayerEvent>,
         renderer: Option<Arc<Mutex<dyn FrameRenderer>>>,
         gl_context: Box<dyn PlayerGLContext>,
-    ) -> Box<dyn Player> {
-        Box::new(player::GStreamerPlayer::new(
-            stream_type,
-            sender,
-            renderer,
-            gl_context,
-        ))
+    ) -> Arc<Mutex<dyn Player>> {
+        let muteable_id = self.next_muteable_id.fetch_add(1, Ordering::Relaxed);
+        let player = player::GStreamerPlayer::new(muteable_id, stream_type, sender, renderer, gl_context);
+        let wrapped_player = Arc::new(Mutex::new(player));
+        let mut muteables = self.muteables.lock().unwrap();
+        let entry = muteables.entry(*id).or_insert(Vec::new());
+        entry.push((muteable_id, Arc::downgrade(&wrapped_player).clone()));
+        wrapped_player
     }
 
-    fn create_audio_context(&self, options: AudioContextOptions) -> AudioContext {
-        AudioContext::new::<Self>(options)
+    fn shutdown_player(
+        &self,
+        id: &ClientContextId,
+        player: Arc<Mutex<dyn Player>>
+    ) {
+        let mut muteables = self.muteables.lock().unwrap();
+        if let Some(vec) = muteables.get_mut(&id) {
+            let player = player.lock().unwrap();
+            let p_id = player.get_id();
+            vec.retain(|m| m.0 != p_id);
+            if vec.len() == 0 {
+                muteables.remove(&id);
+            }
+        }
+
+        if let Err(e) = player.lock().unwrap().shutdown() {
+            warn!("Player was shut down with err: {:?}", e);
+        }
+    }
+
+    fn create_audio_context(
+        &self,
+        id: &ClientContextId,
+        options: AudioContextOptions,
+    ) -> Arc<Mutex<AudioContext>> {
+        let muteable_id = self.next_muteable_id.fetch_add(1, Ordering::Relaxed);
+        let context = Arc::new(Mutex::new(AudioContext::new::<Self>(muteable_id, options)));
+        let mut muteables = self.muteables.lock().unwrap();
+        let entry = muteables.entry(*id).or_insert(Vec::new());
+        entry.push((muteable_id, Arc::downgrade(&context).clone()));
+        context
+    }
+
+    fn shutdown_audio_context(
+        &self,
+        id: &ClientContextId,
+        audio_context: Arc<Mutex<AudioContext>>
+    ) {
+        let mut muteables = self.muteables.lock().unwrap();
+        if let Some(vec) = muteables.get_mut(&id) {
+            let audio_context = audio_context.lock().unwrap();
+            let ac_id = audio_context.get_id();
+            vec.retain(|m| m.0 != ac_id);
+            if vec.len() == 0 {
+                muteables.remove(&id);
+            }
+        }
```

Comment:
Could we move this code to a function (i.e. remove_mutable) and use it from `shutdown_player` and `shutdown_audio_context`?